STM32 RNG: NIST SP800-90B support

gautierg-st opened the pull request:
In order to comply to NIST SP800-90B, the STM32 RNG should set its registers as documented in Application Note AN4230 (Table 3). To that effect, some values have been modified to match the table, and some have been added.

gautierg-st committed:
dts: bindings: rng: stm32: add noise source control property

Some STM32 series need to configure the RNG source noise control register
for NIST SP800-90B compliance.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: entropy: stm32: remove STM32L4 specific code

STM32L4 was using different LL functions from the other series.
With the latest HAL update, the usual functions are now defined, so
the L4-specific code can now be removed.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: arm: st: fix rng nist compliant configs

In order to comply to NIST SP800-90B, the STM32 RNG should set its
registers as documented in Application Note AN4230 (Rev12, Table 3) [1].
To that effect, some values have been modified to match the table, and some
have been added.

For STM32H7R/S and WBA, configure the RNG as per configuration C of the
Reference Manual since the NSCR register is not available yet (internal
issues 221082 and 221083).

[1]: https://www.st.com/resource/en/application_note/dm00073853.pdf

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

mathieuchopstm commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -224,22 +226,36 @@ static void configure_rng(void)
+	    cur_nscr != desired_nscr) {
 		stm32_reg_modify_bits(&rng->CR, cur_nist_cfg, desired_nist_cfg | RNG_CR_CONDRST);
 
+#if defined(CONFIG_SOC_SERIES_STM32L4X)
```
To be sure we clean up this mess in due time:
```suggestion
/*
 * TODO: this is needed only due to mismatch of STM32L4 HAL with others.
 * Remove this special case once STM32CubeL4 v1.18.2 or later has been
 * integrated in hal_stm32 (see also ST Internal Reference 207828).
 */
#if defined(CONFIG_SOC_SERIES_STM32L4X)
```

mathieuchopstm commented on dts/arm/st/l4/stm32l4r5.dtsi:
```diff
@@ -41,5 +41,11 @@
+			/delete-property/ nist-config;
+			/delete-property/ health-test-magic;
+			/delete-property/ health-test-config;
+		};
```
Could use a comment explaining why the values are removed.

gautierg-st commented on dts/arm/st/wba/stm32wba.dtsi:
Yes, and so does the H7R/S series. But these two series lack the NSCR field in the RNG_Typedef definition (as well as the LL functions to set/get the value). See internal tickets 221082 and 221083.

gautierg-st replied:
Since the noise field is not yet available for configuration, I've used Configuration C (from the RefMan) values instead of NIST ones.

mathieuchopstm replied:
A note/comment should be added in the DTSI to clarify why it was done this way; and also, such that the "correct" values can be used once the fields are available in HAL.

mathieuchopstm reviewed (commented):
Otherwise LGTM

etienne-lms reviewed (commented):
I think commit "drivers: entropy: stm32: add noise source control support"
would be worth to be split in 2 commits: first the health test part for L4 series, then the noise source configuration.

etienne-lms reviewed (commented):
For commit "dts: arm: st: fix rng nist compliant configs",
I suggest to explicit the referred application note:
```patch
 (...)
-registers as documented in Application Note AN4230 (Table 3).
+registers as documented in Application Note AN4230 (Rev 12, Table 3) [1].
 (...)
 
+Link: https://www.st.com/resource/en/application_note/dm00073853-stm32-microcontroller-random-number-generation-validation-using-the-nist-statistical-test-suite-stmicroelectronics.pdf [1]
 Signed-off-by: ...
```

gautierg-st committed:
drivers: entropy: stm32: add noise source control support

Set the content of the noise source control register if the corresponding
property is set in device tree.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

erwango reviewed (changes requested):
Otherwise LGTM

gautierg-st commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -220,26 +220,46 @@ static void configure_rng(void)
+#if defined(CONFIG_SOC_SERIES_STM32L4X)
+	cur_htcr = LL_RNG_GetHealthconfiguration(rng);
+#else /* CONFIG_SOC_SERIES_STM32L4X */
 	cur_htcr = LL_RNG_GetHealthConfig(rng);
```
It is fixed in latest L4 HAL release 1.13.6 released a week ago (see https://github.com/STMicroelectronics/stm32l4xx-hal-driver/blob/047a8610c7dc0dcfdb4cd63b86cd3a2cbd3e5899/Inc/stm32l4xx_ll_rng.h#L678).
So I think it's easier to update this HAL and then I'll update this PR.

etienne-lms commented on dts/arm/st/h7rs/stm32h7rs.dtsi:
```diff
@@ -917,6 +917,8 @@
 			clocks = <&rcc STM32_CLOCK(AHB3, 0)>;
 			interrupts = <37 0>;
+			nist-config = <0xf00d>;
+			health-test-config = <0xaac7>;
```
For clarity and maintenance, could you add an inline description to explain these values, as suggested in https://github.com/zephyrproject-rtos/zephyr/pull/99232#discussion_r2518863172?

etienne-lms reviewed (commented):
Did you discard the WBA6x refman config C: `nist-config = <0x820f00>;` in **stm32wba65.dtsi**? It still mentioned in the commit message and would apply.

Note also that **stm32wba.dtsi** sets `nist-config = <0xf00d>;` but the value should be `0xf00d00` (as is recommended in WBA5x refman, config C in its refman) since the implementation does not left shift the DT property value.
